indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java

```diff
@@ -1091,32 +1091,42 @@ private void removeWorker(final Worker worker)
   private void scheduleBlackListedNodesCleanUp()
   {
     ScheduledExecutors.scheduleAtFixedRate(
-            cleanupExec,
-            Period.ZERO.toStandardDuration(),
-            config.getWorkerBlackListCleanupPeriod().toStandardDuration(),
-            new Runnable()
-            {
-              @Override
-              public void run()
-              {
-                long currentTimeStamp = System.currentTimeMillis();
-                for (ZkWorker zkWorker : blackListedWorkers) {
-                  cleanBlackListedNode(zkWorker, currentTimeStamp);
-                }
-              }
+        cleanupExec,
+        Period.ZERO.toStandardDuration(),
+        config.getWorkerBlackListCleanupPeriod().toStandardDuration(),
+        () -> {
+          long currentTimeStamp = System.currentTimeMillis();
+          boolean shouldRunPendingTasks = false;
+
+          // must be synchronized while iterating:
+          // https://docs.oracle.com/javase/8/docs/api/java/util/Collections.html#synchronizedSet-java.util.Set-
+          synchronized (blackListedWorkers) {
+            for (ZkWorker zkWorker : blackListedWorkers) {
+              shouldRunPendingTasks |= cleanBlackListedNode(zkWorker, currentTimeStamp);
             }
+          }
+
+          if (shouldRunPendingTasks) {
+            runPendingTasks();
+          }
+        }
     );
   }
 
-  public void cleanBlackListedNode(ZkWorker zkWorker, long currentTimeStamp)
+  public boolean cleanBlackListedNode(ZkWorker zkWorker, long currentTimeStamp)
   {
-    // Clean blacklisted workers if blacklisted time has elapsed
-    if(currentTimeStamp - zkWorker.getLastCompletedTaskTime().getMillis() >
-            config.getWorkerBlackListBackoffTime().toStandardDuration().getMillis()){
-      // White listing node
-      log.info("Whitelisting worker [%s]. ", zkWorker);
-      blackListedWorkers.remove(zkWorker);
+    long remainingMillis = zkWorker.getLastCompletedTaskTime().getMillis()
```

Comment:
@gianm hm, okay I thought of some behavioral questions along those lines:

- (repeating yours) If a worker gets blacklisted and is running tasks and those tasks later fail, do we consider the blacklist period as starting from when we first decided to blacklist the task (after an earlier failure) or do we start from the latest failure? 
- What if the worker is running tasks and has been blacklisted, but then one of those tasks succeeds? Do we remove it from the blacklist after the success? - Right now this case isn't handled
- After it's removed from the blacklist, does the next failure put it back on the blacklist immediately or do we reset the consecutive failure counter? Right now the behavior is the former and a single subsequent failure re-blacklists the worker.